keras_hub/src/models/basnet/basnet_backbone.py

```diff
@@ -0,0 +1,370 @@
+    for i in reversed(range(num_stages)):
+        if i != (num_stages - 1):  # Except first, scale other decoder stages.
+            x = keras.layers.UpSampling2D(
+                size=2, interpolation="bilinear", dtype=dtype
+            )(x)
+
+        x = keras.layers.concatenate([encoder_blocks[i], x], axis=-1)
+        x = convolution_block(x, filters=filters * 8, dtype=dtype)
+        x = convolution_block(x, filters=filters * 8, dtype=dtype)
+        x = convolution_block(x, filters=filters * 8, dtype=dtype)
+        decoder_blocks.append(x)
+
+    decoder_blocks.reverse()  # Change order from last to first decoder stage.
+    decoder_blocks.append(encoder_blocks[-1])  # Copy bridge to decoder.
+
+    # -------------Side Outputs--------------
+    decoder_blocks = [
+        segmentation_head(decoder_block)  # Prediction segmentation head.
+        for segmentation_head, decoder_block in zip(
+            segmentation_heads, decoder_blocks
+        )
+    ]
+
+    return keras.models.Model(inputs=[x_input], outputs=decoder_blocks)
+
+
+def basnet_rrm(base_model, filters, segmentation_head, dtype=None):
+    """
+    BASNet Residual Refinement Module (RRM).
+
+    This module outputs a fine label map by integrating light encoder,
+    bridge, and decoder blocks.
+
+    Args:
+        base_model: Keras model used as the base or coarse label map.
+        filters: int, the number of filters.
+        segmentation_head: a `keras.layers.Layer`, A Keras layer serving
+            as the segmentation head for refinement module.
+        dtype: `None` or str or `keras.mixed_precision.DTypePolicy`. The dtype
+            to use for the model's computations and weights.
+
+    Returns:
+        A Keras Model that constructs the Residual Refinement Module (RRM).
+    """
+    num_stages = 4
+
+    x_input = base_model.output[0]
+
+    # -------------Encoder--------------
+    x = keras.layers.Conv2D(
+        filters, kernel_size=(3, 3), padding="same", dtype=dtype
+    )(x_input)
+
+    encoder_blocks = []
+    for _ in range(num_stages):
+        x = convolution_block(x, filters=filters)
+        encoder_blocks.append(x)
+        x = keras.layers.MaxPool2D(
+            pool_size=(2, 2), strides=(2, 2), dtype=dtype
+        )(x)
+
+    # -------------Bridge--------------
+    x = convolution_block(x, filters=filters, dtype=dtype)
+
+    # -------------Decoder--------------
+    for i in reversed(range(num_stages)):
+        x = keras.layers.UpSampling2D(
+            size=2, interpolation="bilinear", dtype=dtype
+        )(x)
+        x = keras.layers.concatenate([encoder_blocks[i], x], axis=-1)
+        x = convolution_block(x, filters=filters)
+
+    x = segmentation_head(x)  # Refinement segmentation head.
+
+    # ------------- refined = coarse + residual
+    x = keras.layers.Add(dtype=dtype)(
+        [x_input, x]
+    )  # Add prediction + refinement output
+
+    return keras.models.Model(inputs=base_model.input, outputs=[x])
```

Comment:
newline at end of file (will stop commenting this)

Reply:
I tried to run the shell commands to fix the format issues but i could see `ruff ` command not found as error. I did sync with master before pushing the code to PR but didn't work out the commands. I will figure it out. Thanks